src/inmanta/module.py

```diff
@@ -1205,58 +1442,82 @@ def requires(self) -> "List[Requirement]":
             req = [x for x in parse_requirements(spec)]
             if len(req) > 1:
                 print("Module file for %s has bad line in requirements specification %s" % (self._path, spec))
-            reqe = req[0]
+            reqe = InmantaModuleRequirement(req[0])
             reqs.append(reqe)
         return reqs
 
-    def collect_requirements(self) -> "Mapping[str, Iterable[Requirement]]":
+    def collect_requirements(self) -> "Dict[str, List[InmantaModuleRequirement]]":
         """
-        Collect the list of all requirements of all modules in the project.
+        Collect the list of all module requirements of all modules in the project.
         """
-        specs: Dict[str, List[Requirement]] = {}
+        specs: Dict[str, List[InmantaModuleRequirement]] = {}
         merge_specs(specs, self.requires())
         for module in self.modules.values():
             reqs = module.requires()
             merge_specs(specs, reqs)
         return specs
 
-    def collect_imported_requirements(self) -> "Mapping[str, Iterable[Requirement]]":
+    def collect_imported_requirements(self) -> "Dict[str, List[InmantaModuleRequirement]]":
         imports = set([x.name.split("::")[0] for x in self.get_complete_ast()[0] if isinstance(x, DefineImport)])
         imports.add("std")
-        specs = self.collect_requirements()
+        specs: Dict[str, List[InmantaModuleRequirement]] = self.collect_requirements()
 
-        def get_spec(name: str) -> "Iterable[Requirement]":
+        def get_spec(name: str) -> "List[InmantaModuleRequirement]":
             if name in specs:
                 return specs[name]
-            return parse_requirements(name)
+            return [InmantaModuleRequirement.parse(name)]
 
         return {name: get_spec(name) for name in imports}
 
+    # TODO: test this method
     def verify_requires(self) -> bool:
         """
         Check if all the required modules for this module have been loaded
         """
         LOGGER.info("verifying project")
         imports = set([x.name for x in self.get_complete_ast()[0] if isinstance(x, DefineImport)])
-        modules = self.modules
 
         good = True
 
-        for name, spec in self.collect_requirements().items():
+        for name, module in self.modules.items():
+            if module.GENERATION == ModuleGeneration.V1 and self.module_source.path_for(name) is not None:
+                LOGGER.warning(
+                    "Module %s was installed installed as v1 but later pulled in as v2 a dependency as v2."
+                    " Continuing would lead to inconsistent behavior. A rerun should resolve this issue.",
+                    name,
+                )
+                good = False
+
+        requirements: Dict[str, List[InmantaModuleRequirement]] = self.collect_requirements()
```

Comment:
Would it not be simpler to 
- have every module know its own version 
- every module can return its module requirements/constraints
- evaluate the constraints against the set of loaded versions?

Reply:
refactored the whole verify logic to
1. accomodate this
2. verify the modules cache